File: algo/leetcode_122.py
''' Best time to buy and sell stocks (unlimited transactions)
'''

def maxProfit(prices):
    ''' This assumes that you can buy-sell "on the same day"
    '''
    if not prices: return 0
    n = len(prices)
    if n < 2: return 0
    
    profit = 0
    for i in range(n - 1):
        diff = prices[i+1] - prices[i]
        if diff > 0:
            profit += diff

    return profit

def maxProfitMinTrans(prices):
    ''' This actually gives a solution with the minimum 
        number of transactions need to achieve max profit
    '''
    if not prices: return 0
    n = len(prices)
    if n < 2: return 0
    i = 0    
    
    # Find the first local minimum
    while i + 1 < n:
        if prices[i] < prices[i+1]:
            #print('BUY', prices[i])
            break
        i += 1
    hold = prices[i]; i += 1
    
    profit = 0
    while i + 1 < n:
        # Find local maximum for sell
        while i + 1 < n:
            a, b, c = prices[i-1], prices[i], prices[i+1]
            if a <= b and b >= c and not (a == b and b == c):
                profit += prices[i] - hold
                #print('SELL', prices[i])
                hold = None
                break
            i += 1
        
        # Find local minimum to buy
        while i + 1 < n:
            a, b, c = prices[i-1], prices[i], prices[i+1]        
            if a >= b and b <= c and not (a == b and b == c):
                hold = prices[i]
                #print('BUY', hold)
                break
            i += 1
            
    if hold is not None and i < n:
        profit += prices[i] - hold
    
    return profit


def TEST(prices):
    print('max profit multiple transactions:', 
          maxProfit(prices))

TEST([7, 1, 5, 3, 6, 4])

File: algo/leetcode_122.cxx
#include <iostream>
#include <vector>

using namespace std;

int maxProfit(vector<int>& prices) {
    if (prices.size() < 2) return 0;

    int profit = 0;    
    int curr_hold = -1;
    int idx = 0;
    while (true) {
        for (; idx + 1 < prices.size(); ++idx) {
            if (prices[idx] < prices[idx+1]) {
                curr_hold = prices[idx];
                cout << "BUY " << prices[idx] << endl;
                ++idx; break;
            }
        }
        for (; idx + 1 < prices.size(); ++idx) {
            if (prices[idx] > prices[idx + 1]) {
                profit += prices[idx] - curr_hold;
                cout << "SEL " << prices[idx] << endl;
                curr_hold = -1;
                ++idx; break;
            }
        }
        if (idx + 1 == prices.size()) {
            if (-1 != curr_hold) {
                cout << "SEL " << prices[idx] << endl;
                profit += prices[idx] - curr_hold;
            }
            break;
        }
    }    
    return profit;
}

void TEST(vector<int> prices, int tgt) {    
    int res = maxProfit(prices);
    if (res != tgt)
        cout << "RES " << res << " != TGT " << tgt << endl; 
    else 
        cout << "OK" << endl;
}

int main() {
    TEST({8,4,1,2,3,3,2,1,5,6,7,2,3,1,2,2}, 10);
    TEST({2,1,2,0,1}, 2);
    TEST({2,1}, 0);
    TEST({1,2}, 1);
}
